lib/bedrock/index.js

```diff
@@ -27,6 +43,54 @@ module.exports = (data) => {
       }
 
       return itemstates
+    },
+
+    loadCustomBlockStates (blockStates) {
+      const blocks = []
+      for (const block of blockStates) {
+        let [ns, name] = block.name.split(':')
+        if (!name) {
+          name = ns
+          ns = 'minecraft'
+        }
+        blocks.push({ name: `${ns}:${name}`, runtime_id: block.state.value.vanilla_block_data.value.block_id.value, component_based: ns !== 'minecraft' })
+      }
+      data.blockStates.push(...blockStates)
+      data.blocksArray.push(...blocks)
+      data.blocks = buildIndexFromArray(data.blocksArray, 'id')
+      data.blocksByName = buildIndexFromArray(data.blocksArray, 'name')
+    },
+
+    writeCustomBlockStates () {
+      const blockstates = []
+      for (const block of data.blocksArray) {
+        // Custom items with different namespaces can also be in the palette
+        let [ns, name] = block.name.split(':')
+        if (!name) {
+          name = ns
+          ns = 'minecraft'
+        }
+
+        blockstates.push({ name: `${ns}:${name}`, runtime_id: block.state.value.vanilla_block_data.value.block_id.value, component_based: ns !== 'minecraft' })
+      }
+
+      return blockstates
+    },
+
+    calculateBlockStateHashes () {
+      const pBlock = require('prismarine-block')(this, this.version)
+      const blocksByRuntimeId = {}
+      for (const [stateId, block] of Object.entries(data.blockStates)) {
+        const name = `minecraft:${block.name}`
+        const states = {}
+        for (const key of Object.keys(block.states).sort()) {
+          states[key] = block.states[key]
+        }
```

Comment:
Should be moved to pblock

Reply:
Should that be a field or a function? Should it be only avalieble if blockHashes supported?

Reply:
I think i understand, keys should be pre-sorted in pblock

Reply:
I don't understand the question. All prismarine-registry should be doing is 
```js
  function loadHashedRuntimeIds () {
    const Block = require('prismarine-block')(this)
    data.blocksByRuntimeId = {}
    for (let i = 0; i < data.blockStates.length; i++) {
      const { name, states } = data.blockStates[i]
      const hash = Block.getHash(name, states)
      data.blocksByRuntimeId[hash] = data.blocksByStateId[i]
    }
  }
```

Reply:
BlocksByState have stateId excluded, should indexers keep that field/new indexer that keeps indexing field? Or technically since its sequencial index of hash in runtimeIds would give stateId?

Reply:
In JavaScript, objects are passed by reference, not by copy. This means when you mutate one reference (like adding a .stateId to it) then all other references will also be updated. This effectively means iterating over blocksByStateId or any other blocks array and mutating a block object in one of them will update it everywhere else.

We don't want updates in prismarine-registry to blocks to affect the static data exposed by node-minecraft-protocol, so an explicit copy would need to be made before any edits. That's what I was mentioning in the prismarine-chunk comment